File: csharp/tests/TestSpecialErrorDeser.cs
using System.Text.Json;
using MyAirtable;
using Xunit;

namespace MyAirtable.Tests;

/// <summary>
/// CS5.7 (+ CS5.3 end-to-end) — parity with rust test_special_error_deser / Kotlin/Swift/Java
/// TestSpecialErrorDeser. Offline decoding of Airtable's <c>{"specialValue": ...}</c> /
/// <c>{"error": ...}</c> computed-field shapes through the generated module, plus a generated-model
/// decode proving erroring/special formula + lookup/rollup fields don't break deserialization.
/// </summary>
public class TestSpecialErrorDeser
{
    private static T Decode<T>(string raw) =>
        JsonSerializer.Deserialize<T>(raw, AirtableJson.Options)!;

    [Fact]
    public void SpecialNumberAcceptsObjectFormRejectsArrayForm()
    {
        Assert.Equal("NaN", Decode<SpecialNumber>("""{"specialValue":"NaN"}""").SpecialValue);
        Assert.ThrowsAny<Exception>(() => Decode<SpecialNumber>("""["NaN"]"""));
    }

    [Fact]
    public void ErrorValueAcceptsObjectFormRejectsArrayForm()
    {
        Assert.Equal("#ERROR!", Decode<ErrorValue>("""{"error":"#ERROR!"}""").Error);
        Assert.ThrowsAny<Exception>(() => Decode<ErrorValue>("""["#ERROR!"]"""));
    }

    [Fact]
    public void MaybeSpecialOrErrorStringParsesPlainStringRejectsLookupArray()
    {
        var plain = Decode<MaybeSpecialOrError<string>>("\"Stukenholtz\"");
        Assert.Equal("Stukenholtz", plain.ValueOrDefault);
        // Pre-fix regression in other targets: ["Stukenholtz"] accidentally succeeded as Special.
        Assert.ThrowsAny<Exception>(() =>
            Decode<MaybeSpecialOrError<string>>("""["Stukenholtz"]""")
        );
    }

    [Fact]
    public void VecOrValueLookupShapes()
    {
        var single = Decode<VecOrValue<MaybeSpecialOrError<string>>>("\"hello\"");
        var singleShape = Assert.IsType<VecOrValue<MaybeSpecialOrError<string>>.Single>(single);
        Assert.Equal("hello", singleShape.Value!.ValueOrDefault);

        var multiple = Decode<VecOrValue<MaybeSpecialOrError<string>>>(
            """["Stukenholtz Laboratory Inc"]"""
        );
        Assert.Equal(
            new List<string> { "Stukenholtz Laboratory Inc" },
            VecOrValue.CleanValues(multiple)
        );

        var withNulls = Decode<VecOrValue<MaybeSpecialOrError<string>>>("""["a", null, "b"]""");
        var multipleShape = Assert.IsType<VecOrValue<MaybeSpecialOrError<string>>.Multiple>(
            withNulls
        );
        Assert.Equal(3, multipleShape.Values.Count);
        Assert.Null(multipleShape.Values[1]);

        var special = Decode<VecOrValue<MaybeSpecialOrError<double>>>("""{"specialValue":"NaN"}""");
        var specialValue = Assert.IsType<MaybeSpecialOrError<double>.Special>(
            Assert.IsType<VecOrValue<MaybeSpecialOrError<double>>.Single>(special).Value
        );
        Assert.Equal("NaN", specialValue.Number.SpecialValue);

        var error = Decode<VecOrValue<MaybeSpecialOrError<double>>>("""{"error":"#ERROR!"}""");
        var errorValue = Assert.IsType<MaybeSpecialOrError<double>.Error>(
            Assert.IsType<VecOrValue<MaybeSpecialOrError<double>>.Single>(error).Value
        );
        Assert.Equal("#ERROR!", errorValue.Err.Error);
    }

    [Fact]
    public void ModelDecodesErroringComputedFields()
    {
        // The exact failure mode that used to throw in other targets: Airtable returns
        // error/special objects for computed fields, while lookup returns an array with nulls.
        var fieldsJson = $$"""
            {
              "{{PrimaryFields.PrimaryKeyId}}": "ErrDeser",
              "{{PrimaryFields.FormulaSimpleId}}": {"error": "#ERROR!"},
              "{{PrimaryFields.FormulaIdId}}": "recERRDESER123456",
              "{{PrimaryFields.LookupId}}": ["Lab A", null],
              "{{PrimaryFields.RollupId}}": {"specialValue": "NaN"}
            }
            """;
        var model = JsonSerializer.Deserialize<PrimaryModel>(fieldsJson, AirtableJson.Options)!;

        Assert.Equal("ErrDeser", model.PrimaryKey);
        // Error variant — ValueOrDefault is the non-nullable double default (0) for value-type T,
        // so the variant type IS the assertion.
        Assert.IsType<MaybeSpecialOrError<double>.Error>(model.FormulaSimple);
        Assert.Equal("recERRDESER123456", model.FormulaId!.ValueOrDefault);
        Assert.Equal(new List<string> { "Lab A" }, VecOrValue.CleanValues(model.Lookup));
        var lookupShape = Assert.IsType<VecOrValue<MaybeSpecialOrError<string>>.Multiple>(
            model.Lookup
        );
        Assert.Equal(2, lookupShape.Values.Count);
        var rollupSpecial = Assert.IsType<MaybeSpecialOrError<string>.Special>(
            Assert.IsType<VecOrValue<MaybeSpecialOrError<string>>.Single>(model.Rollup).Value
        );
        Assert.Equal("NaN", rollupSpecial.Number.SpecialValue);
    }
}

File: cpp/tests/test_special_error_deser.cpp
// Offline decoding of Airtable's {"specialValue": ...} / {"error": ...}
// computed-field shapes through the generated module, plus a generated-model
// decode proving erroring/special formula + lookup/rollup fields don't break
// deserialization (C# TestSpecialErrorDeser parity, 5 cases).
#include <catch2/catch_test_macros.hpp>

#include <string>
#include <vector>

#include "test_setup.hpp"

using namespace myairtable;

namespace {

template <typename T> T decode(const std::string& raw) {
    return json::parse(raw).get<T>();
}

} // namespace

TEST_CASE("special error deser: special number accepts object form rejects array form",
          "[json][special-error-deser]") {
    REQUIRE(decode<SpecialNumber>(R"({"specialValue":"NaN"})").special_value == "NaN");
    REQUIRE_THROWS(decode<SpecialNumber>(R"(["NaN"])"));
}

TEST_CASE("special error deser: error value accepts object form rejects array form",
          "[json][special-error-deser]") {
    REQUIRE(decode<ErrorValue>(R"({"error":"#ERROR!"})").error == "#ERROR!");
    REQUIRE_THROWS(decode<ErrorValue>(R"(["#ERROR!"])"));
}

TEST_CASE("special error deser: maybe-special-or-error string parses plain string rejects "
          "lookup array",
          "[json][special-error-deser]") {
    const auto plain = decode<MaybeSpecialOrError<std::string>>(R"("Stukenholtz")");
    REQUIRE(plain.value() == "Stukenholtz");
    // Pre-fix regression in other targets: ["Stukenholtz"] accidentally
    // succeeded as Special.
    REQUIRE_THROWS(decode<MaybeSpecialOrError<std::string>>(R"(["Stukenholtz"])"));
}

TEST_CASE("special error deser: vec-or-value lookup shapes", "[json][special-error-deser]") {
    const auto single = decode<VecOrValue<MaybeSpecialOrError<std::string>>>(R"("hello")");
    REQUIRE(single.is_single());
    REQUIRE(single.value()->value() == "hello");

    const auto multiple =
        decode<VecOrValue<MaybeSpecialOrError<std::string>>>(R"(["Stukenholtz Laboratory Inc"])");
    REQUIRE(multiple.clean_values() == std::vector<std::string>{"Stukenholtz Laboratory Inc"});

    const auto with_nulls =
        decode<VecOrValue<MaybeSpecialOrError<std::string>>>(R"(["a", null, "b"])");
    REQUIRE(with_nulls.is_multiple());
    REQUIRE(with_nulls.values().size() == 3);
    REQUIRE(with_nulls.values()[1] == std::nullopt);

    const auto special =
        decode<VecOrValue<MaybeSpecialOrError<double>>>(R"({"specialValue":"NaN"})");
    REQUIRE(special.is_single());
    REQUIRE(special.value()->is_special());
    REQUIRE(special.value()->special()->special_value == "NaN");

    const auto error = decode<VecOrValue<MaybeSpecialOrError<double>>>(R"({"error":"#ERROR!"})");
    REQUIRE(error.is_single());
    REQUIRE(error.value()->is_error());
    REQUIRE(error.value()->error()->error == "#ERROR!");
}

TEST_CASE("special error deser: model decodes erroring computed fields",
          "[json][special-error-deser]") {
    // The exact failure mode that used to throw in other targets: Airtable
    // returns error/special objects for computed fields, while lookup returns
    // an array with nulls.
    const json fields = {
        {std::string(PrimaryFields::kPrimaryKeyId), "ErrDeser"},
        {std::string(PrimaryFields::kFormulaSimpleId), {{"error", "#ERROR!"}}},
        {std::string(PrimaryFields::kFormulaIdId), "recERRDESER123456"},
        {std::string(PrimaryFields::kLookupId), {"Lab A", nullptr}},
        {std::string(PrimaryFields::kRollupId), {{"specialValue", "NaN"}}},
    };
    const json record = {{"fields", fields}};
    const auto model = record.get<PrimaryModel>();

    REQUIRE(model.primary_key == "ErrDeser");
    // Error variant — the variant type IS the assertion.
    REQUIRE(model.formula_simple.has_value());
    REQUIRE(model.formula_simple->is_error());
    REQUIRE(model.formula_id.has_value());
    REQUIRE(model.formula_id->value() == "recERRDESER123456");
    REQUIRE(model.lookup.has_value());
    REQUIRE(model.lookup->clean_values() == std::vector<std::string>{"Lab A"});
    REQUIRE(model.lookup->is_multiple());
    REQUIRE(model.lookup->values().size() == 2);
    REQUIRE(model.rollup.has_value());
    REQUIRE(model.rollup->is_single());
    REQUIRE(model.rollup->value()->is_special());
    REQUIRE(model.rollup->value()->special()->special_value == "NaN");
}
